File: Example/ExampleObjectiveC/PIONotificationViewController.m
//
//  PIONotificationViewController.m
//  ExampleObjectiveC
//
//  Created by zee-pk on 26/08/2016.
//  Copyright (c) 2016 predict.io by ParkTAG GmbH. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "PIONotificationViewController.h"
#import "AppDelegate.h"
#import "EventViaNotification.h"
#import "PIOMapViewController.h"

@interface PIONotificationViewController ()

@property (strong, nonatomic) PredictIOService *predictIOService;
@property (strong, nonatomic) NSDateFormatter *dateFormatter;
@property (strong, nonatomic) NSArray *transportationModeLabels;
@property (strong, nonatomic) NSArray *labels;
@property (strong, nonatomic) NSFetchedResultsController *fetchedResultsController;
@property (strong, nonatomic) NSManagedObjectContext *managedObjectContext;

@end

@implementation PIONotificationViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    AppDelegate *appDelegate = (AppDelegate *) [UIApplication sharedApplication].delegate;
    self.predictIOService = appDelegate.predictIOService;
    self.managedObjectContext = appDelegate.managedObjectContext;
    
    self.dateFormatter = [[NSDateFormatter alloc] init];
    self.dateFormatter.dateFormat = @"dd/MM hh:mm a";
    self.labels = @[@"Departing", @"Departed", @"Departure Cancel", @"STMP Callback", @"Arrival Suspected", @"Arrived", @"Searching in perimeter"];
    self.transportationModeLabels = @[@"TransportationMode: Undetermined", @"TransportationMode: Car", @"TransportationMode: NonCar"];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    BOOL predictIOEnabled = [defaults boolForKey:@"PredictIOEnabled"];
    self.navigationItem.rightBarButtonItem.title = predictIOEnabled ? @"Stop" : @"Start";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"showOnMap"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        EventViaNotification *event = (EventViaNotification *) [[self fetchedResultsController] objectAtIndexPath:indexPath];
        CLLocation *location = [[CLLocation alloc] initWithCoordinate:CLLocationCoordinate2DMake(event.latitude.doubleValue, event.longitude.doubleValue) altitude:0.0 horizontalAccuracy:event.accuracy.doubleValue verticalAccuracy:0.0 course:0.0 speed:0.0 timestamp:event.timeStamp];
        PIOMapViewController *controller = [segue destinationViewController];
        controller.location = location;
    }
}

- (IBAction)startStopPredictIO:(id)sender {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    BOOL predictIOEnabled = [defaults boolForKey:@"PredictIOEnabled"];
    if (predictIOEnabled) {
        self.navigationItem.rightBarButtonItem.title = @"Start";
        [self.predictIOService stop];
    } else {
        self.navigationItem.rightBarButtonItem.title = @"Stop";
        [self.predictIOService startWithCompletionHandler:^(NSError *error) {
            if (error) {
                [NSOperationQueue.mainQueue addOperationWithBlock:^{
                    NSString *errorTitle = error.userInfo[@"NSLocalizedFailureReason"];
                    NSString *errorDescription = error.userInfo[@"NSLocalizedDescription"];
                    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:errorTitle
                                                                                             message:errorDescription
                                                                                      preferredStyle:UIAlertControllerStyleAlert];
                    UIAlertAction *alertActionOK = [UIAlertAction
                                                    actionWithTitle:@"OK"
                                                    style:UIAlertActionStyleDefault
                                                    handler:nil];
                    [alertController addAction:alertActionOK];
                    [self presentViewController:alertController animated:YES completion:nil];
                    self.navigationItem.rightBarButtonItem.title = @"Start";
                    NSLog(@"<predict.io> API Key: %@ %@", errorTitle, errorDescription);
                    [defaults setBool:NO forKey:@"PredictIOEnabled"];
                    [defaults synchronize];
                }];
            } else {
                NSLog(@"Started predict.io...");
            }
        }];
    }
}

#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.fetchedResultsController.sections.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> sectionInfo = self.fetchedResultsController.sections[(NSUInteger) section];
    return sectionInfo.numberOfObjects;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    NSDictionary *infoDictionary = [NSBundle mainBundle].infoDictionary;
    NSString *build = infoDictionary[@"CFBundleVersion"];
    NSString *version = [PredictIO sharedInstance].version;
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"HeaderCell"];
    cell.textLabel.text = [NSString stringWithFormat:@"SDK %@", version];
    cell.detailTextLabel.text = [NSString stringWithFormat:@"Build %@", build];
    
    return cell;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    NSManagedObject *object = [[self fetchedResultsController] objectAtIndexPath:indexPath];
    [self configureCell:cell withEvent:(EventViaNotification *)object];
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return NO;
}

- (void)configureCell:(UITableViewCell *)cell withEvent:(EventViaNotification *)event {
    enum PredictIOEventType eventType = (enum PredictIOEventType) event.type.integerValue;
    if (eventType == TransportMode) {
        cell.textLabel.text = self.transportationModeLabels[(NSUInteger) event.mode.integerValue];
    } else {
        cell.textLabel.text = self.labels[eventType];
    }
    cell.detailTextLabel.text = [self.dateFormatter stringFromDate:event.timeStamp];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self performSegueWithIdentifier:@"showOnMap" sender:self];
}

#pragma mark - Fetched results controller

- (NSFetchedResultsController *)fetchedResultsController {
    if (_fetchedResultsController != nil) {
        return _fetchedResultsController;
    }

    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
    // Edit the entity name as appropriate.
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"EventViaNotification" inManagedObjectContext:self.managedObjectContext];
    [fetchRequest setEntity:entity];

    // Set the batch size to a suitable number.
    [fetchRequest setFetchBatchSize:20];

    // Edit the sort key as appropriate.
    NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"timeStamp" ascending:NO];

    [fetchRequest setSortDescriptors:@[sortDescriptor]];

    // Edit the section name key path and cache name if appropriate.
    // nil for section name key path means "no sections".
    NSFetchedResultsController *aFetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:self.managedObjectContext sectionNameKeyPath:nil cacheName:@"Master"];
    aFetchedResultsController.delegate = self;
    self.fetchedResultsController = aFetchedResultsController;

    NSError *error = nil;
    if (![self.fetchedResultsController performFetch:&error]) {
        // Replace this implementation with code to handle the error appropriately.
        // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
        NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
        abort();
    }

    return _fetchedResultsController;
}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}

- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type {
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;

        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;

        default:
            return;
    }
}

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath {
    UITableView *tableView = self.tableView;

    switch(type) {
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:@[newIndexPath] withRowAnimation:UITableViewRowAnimationFade];
            break;

        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
            break;

        case NSFetchedResultsChangeUpdate:
            [self configureCell:[tableView cellForRowAtIndexPath:indexPath] withEvent:(EventViaNotification *)anObject];
            break;

        case NSFetchedResultsChangeMove:
            [tableView moveRowAtIndexPath:indexPath toIndexPath:newIndexPath];
            break;
    }
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

@end

File: ExampleSwift/ExampleSwift/PIONotificationViewController.swift
//
//  PIONotificationViewController.swift
//  ExampleSwift
//
//  Created by zee-pk on 23/08/2016.
//  Copyright (c) 2016 predict.io by ParkTAG GmbH. All rights reserved.
//

import UIKit
import CoreData
import Foundation
import PredictIO

class PIONotificationViewController: UITableViewController, NSFetchedResultsControllerDelegate {
    
    let dateFormatter = DateFormatter()
    
    let labels = ["Departing", "Departed", "Departure Cancelled", "STMP Callback", "Arrival Suspected", "Arrived", "Searching in perimeter", "Stationary after arrival", "TraveledByAirPlane"]
    
    let transportationModeLabels = ["TransportationMode: Undetermined", "TransportationMode: Car", "TransportationMode: NonCar"];
    
    let stationaryStates = ["Stationary: NO", "Stationary: YES"]

    let managedObjectContext: NSManagedObjectContext = {
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        return appDelegate.managedObjectContext
    }()
    
    let predictIOService: PredictIOService = {
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        return appDelegate.predictIOService
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        dateFormatter.dateFormat = "dd/MM hh:mm a"
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let defaults = UserDefaults.standard
        let predictIOEnabled = defaults.bool(forKey: "PredictIOEnabled")
        navigationItem.rightBarButtonItem!.title = predictIOEnabled ? "Stop" : "Start"
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func startStopPredictIO(_ sender: AnyObject) {
        let defaults = UserDefaults.standard
        let predictIOEnabled = defaults.bool(forKey: "PredictIOEnabled")
        if predictIOEnabled {
            navigationItem.rightBarButtonItem!.title = "Start"
            predictIOService.stop()
        } else {
            navigationItem.rightBarButtonItem!.title = "Stop"
            predictIOService.startWithCompletionHandler({ (error) -> (Void) in
                if (error != nil) {
                    OperationQueue.main.addOperation({
                        let userInfo = error!._userInfo as! NSDictionary
                        let errorTitle = userInfo["NSLocalizedFailureReason"] as! String
                        let errorDescription = userInfo["NSLocalizedDescription"] as! String
                        let alertController = UIAlertController(title: errorTitle, message: errorDescription, preferredStyle: .alert)
                        let alertActionOK = UIAlertAction(title: "OK", style: .default, handler: nil)
                        alertController.addAction(alertActionOK)
                        self.present(alertController, animated: true, completion: nil)
                        self.navigationItem.rightBarButtonItem!.title = "Start"
                        print("<predict.io> API Key: \(errorTitle) \(errorDescription)")
                        defaults.set(false, forKey: "PredictIOEnabled")
                        defaults.synchronize()
                    })
                } else {
                    print("Started predict.io...")
                }
            })
        }
    }
    
    @IBAction func showHomeWorkZones(_ sender: Any) {
        let homeZone = PredictIO.sharedInstance().homeZone
        let workZone = PredictIO.sharedInstance().workZone
        
        if ((homeZone != nil) || (workZone != nil)) {
            performSegue(withIdentifier: "showZones", sender: self)
        } else {
            let alertController = UIAlertController(title: "Home/Work Zones", message: "No zone information available at this moment. Please check again after some trips.", preferredStyle: .alert)
            let alertActionOK = UIAlertAction(title: "OK", style: .default, handler: nil)
            alertController.addAction(alertActionOK)
            present(alertController, animated: true, completion: nil)
        }
    }
    // MARK: - Table View
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return fetchedResultsController.sections?.count ?? 0
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo = fetchedResultsController.sections![section]
        return sectionInfo.numberOfObjects
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let infoDictionary = Bundle.main.infoDictionary
        let build = infoDictionary!["CFBundleVersion"];
        let version = PredictIO.sharedInstance().version
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "HeaderCell")
        cell?.textLabel?.text = "SDK v\(version!)"
        cell?.detailTextLabel?.text = "Build v\(build!)"
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let event = fetchedResultsController.object(at: indexPath)
        configureCell(cell, withEvent: event)
        return cell
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return false
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        performSegue(withIdentifier: "showOnMap", sender: self)
    }

    func configureCell(_ cell: UITableViewCell, withEvent event: EventViaNotification) {
        let eventTypeIntegerValue = event.type!.intValue;
        let eventType = PredictIOEventType(rawValue: eventTypeIntegerValue)
        if (eventType == .transportMode) {
            let modeIntegerValue = event.mode!.intValue
            cell.textLabel!.text = transportationModeLabels[modeIntegerValue]
        } else if (eventType == .stationary) {
            let stationaryIntegerValue = event.stationary!.intValue
            cell.textLabel!.text = stationaryStates[stationaryIntegerValue]
        } else {
            cell.textLabel!.text = labels[eventTypeIntegerValue]
        }
        cell.detailTextLabel!.text =  dateFormatter.string(from: event.timeStamp! as Date)
    }
    
    // MARK: - Navigation

    override func prepare(for segue: UIStoryboardSegue, sender: Any!) {
        if (segue.identifier == "showOnMap") {
            let indexPath = tableView.indexPathForSelectedRow!
            let event = fetchedResultsController.object(at: indexPath)
            let coordinate = CLLocationCoordinate2DMake(event.latitude!.doubleValue, event.longitude!.doubleValue)
            let location = CLLocation(coordinate: coordinate, altitude: 0.0, horizontalAccuracy: event.accuracy!.doubleValue , verticalAccuracy: 0.0, course: 0.0, speed: 0.0, timestamp: event.timeStamp! as Date)
            let controller = segue.destination as! PIOMapViewController
            controller.location = location
            controller.zoneType = PIOZoneType(rawValue: event.zoneType as! Int32)!
        } else if (segue.identifier == "showZones") {
            let controller = segue.destination as! PIOZoneViewController
            controller.homeZone = PredictIO.sharedInstance().homeZone
            controller.workZone = PredictIO.sharedInstance().workZone
        }
    }

    // MARK: - Fetched results controller
    
    var fetchedResultsController: NSFetchedResultsController<EventViaNotification> {
        if _fetchedResultsController != nil {
            return _fetchedResultsController!
        }
        
        let fetchRequest: NSFetchRequest<EventViaNotification> = EventViaNotification.fetchRequest()
        
        // Set the batch size to a suitable number.
        fetchRequest.fetchBatchSize = 20
        
        // Edit the sort key as appropriate.
        let sortDescriptor = NSSortDescriptor(key: "timeStamp", ascending: false)
        
        fetchRequest.sortDescriptors = [sortDescriptor]
        
        // Edit the section name key path and cache name if appropriate.
        // nil for section name key path means "no sections".
        let aFetchedResultsController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: self.managedObjectContext, sectionNameKeyPath: nil, cacheName: nil)
        aFetchedResultsController.delegate = self
        _fetchedResultsController = aFetchedResultsController
        
        do {
            try _fetchedResultsController!.performFetch()
        } catch {
            // Replace this implementation with code to handle the error appropriately.
            // fatalError() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
            let nserror = error as NSError
            fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
        }
        
        return _fetchedResultsController!
    }
    var _fetchedResultsController: NSFetchedResultsController<EventViaNotification>? = nil
    
    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.beginUpdates()
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange sectionInfo: NSFetchedResultsSectionInfo, atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType) {
        switch type {
        case .insert:
            tableView.insertSections(IndexSet(integer: sectionIndex), with: .fade)
        case .delete:
            tableView.deleteSections(IndexSet(integer: sectionIndex), with: .fade)
        default:
            return
        }
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        switch type {
        case .insert:
            tableView.insertRows(at: [newIndexPath!], with: .fade)
        case .delete:
            tableView.deleteRows(at: [indexPath!], with: .fade)
        case .update:
            configureCell(tableView.cellForRow(at: indexPath!)!, withEvent: anObject as! EventViaNotification)
        case .move:
            tableView.moveRow(at: indexPath!, to: newIndexPath!)
        }
    }
    
    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.endUpdates()
    }
}
